src/HotChocolate/Core/src/Execution/Pipeline/OperationExecutionMiddleware.cs

```diff
@@ -97,8 +98,19 @@ internal sealed class OperationExecutionMiddleware
                 clonedContext.Services.GetRequiredService<DefaultRequestContextAccessor>();
             accessor.RequestContext = clonedContext;
 
+            // This prevents a closure from being formed over the clonedContext making the schema long lived.
```

Comment:
So we have the memory leak in combination with subscriptions? Is that because subscriptions somehow survive a schema phase out?

Reply:
Correct, I would categorise this as three issues.
1. Eviction did not lead to correct disposal of the schema due to the reversed dispose check.
2. Attempting to access a disposed ServiceProvider (https://github.com/ChilliCream/hotchocolate/blob/mp/schema-memory-pressure/src/HotChocolate/Stitching/src/Stitching.Redis/RedisExecutorOptionsProvider.cs#L109) during the eviction actions (https://github.com/ChilliCream/hotchocolate/blob/main/src/HotChocolate/Core/src/Execution/DependencyInjection/RequestExecutorServiceCollectionExtensions.cs#L143)
3. The cloned request context (https://github.com/ChilliCream/hotchocolate/blob/mp/schema-memory-pressure/src/HotChocolate/Core/src/Execution/Pipeline/OperationExecutionMiddleware.cs#L95) is a shallow clone, and when added to the closure, it causes additional schema versions to be retained for the lifetime of the running Websocket HttpContext.

Reply:
OK, I kind of understand the issue now ... Still, this will introduce a new memory leak since we are not handling the ServiceProvider itself.

```csharp
await using ServiceProvider services = new ServiceCollection()
```

If we do not dispose of this we still will have a leakage.

The main issue is how we extract the options I think.

Reply:
Disposal of the ServiceProvider does occur from here: https://github.com/ChilliCream/hotchocolate/blob/mp/schema-memory-pressure/src/HotChocolate/Core/src/Execution/RequestExecutorResolver.cs#L185

Then: https://github.com/ChilliCream/hotchocolate/blob/mp/schema-memory-pressure/src/HotChocolate/Core/src/Execution/RequestExecutorResolver.cs#L483

Reply:
Of course, happy to discuss further refactoring opportunities for this. I was looking for surgical change rather than more fundamental ones :)

Reply:
You are mistaken there. We have two `ServiceProvider`. The outer-one on which we do `AddGraphQL` and the inner one that we use to register our schema internal services on. This is done so that we can host multiple executors in the same root service provider. When you dispose the executor you will only dispose the inner (executor bound) ServiceProvider.

Reply:
Further, as far as I remember, we create this ServiceProvider in the RedisProvider only to extract the options. It's kind of wrong that something is captured there. In this case, we are actually dealing with three ServiceProvider. One that is the actual ASP.Net Core service provider, the temporary one that is used to extract the options, and another one for the executor.

Reply:
Ah awesome! Thanks for that Michael. The redis SP is used during eviction actions which was part of the original problem and why I was trying to move the disposal to the executor disposal. So I'm not immediately sure on the right approach to resolve this. Any suggestions?

Reply:
Not sure either. Let's keep this a blocked problem. I think we can solve the while issue simpler with the refactoring of the stitching engine. I will keep the PR open and mark it as blocked so we have it as a reminder.